### ISSUE TITLE
Is testmode broken?

### ISSUE THREAD

**Issue by unhandyandy** [FluxML/Flux.jl#333]
Here's a mwe.  I create a simple chain with BN, then I follow the instructions at https://github.com/FluxML/Flux.jl/issues/232
to create a predict function in testmode.

```
julia> testchain = Chain(Dense(10,10,relu,initb=Flux.glorot_normal),BatchNorm(10))
Chain(Dense(10, 10, NNlib.relu), BatchNorm(10))

julia> params(testchain)
4-element Array{Any,1}:
 param([0.209142 0.241711 … -0.523774 0.260564; 0.246958 -0.389062 … 0.0435414 -0.361227; … ; -0.511417 -0.124041 … -0.324386 0.407871; -0.0613081 0.271772 … -0.223136 0.352242])
 param([-0.0894909, 0.335661, -0.1677, -0.454554, 0.207357, -0.214411, -0.326957, -0.713971, -0.442211, 0.517471])                                                                
 param([0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0])                                                                                                                        
 param([1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0])                      

julia> testpred = mapleaves(Flux.Tracker.data, testchain)
Chain(Dense(10, 10, NNlib.relu), BatchNorm(10))

julia> Flux.testmode!(testpred)
Chain(Dense(10, 10, NNlib.relu), BatchNorm(10))
```

Now I feed some random data to predict.

```
julia> testpred(rand(10,1))
10×1 Array{Float64,2}:
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0

julia> testpred(rand(10,1))
10×1 Array{Float64,2}:
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0

```
There's no way I should be getting all zeros if testmode is working correctly.

Right?  Or what did I do wrong?

**Comment by unhandyandy** [comment]
After restarting Julia this problem no longer occurs.  So the problem seems to be some kind of corruption?

**Comment by ViralBShah** [comment]
Corruption is a bug. Ideally we want to find it and fix it.

**Comment by unhandyandy** [comment]
Here's a fresh run of Julia.  I define the testchain and testpred function as above.  If I call testpred first before putting it into testmode, it seems I am unable to do it later.

julia> using Flux

julia> testchain = Chain(Dense(10,10,relu,initb=Flux.glorot_normal),BatchNorm(10))
Chain(Dense(10, 10, NNlib.relu), BatchNorm(10))

julia> params(testchain)
4-element Array{Any,1}:
 param([0.43428 0.419424 … -0.257435 0.165523; 0.169282 -0.0603898 … 0.459877 -0.228518; … ; 0.0777249 0.136552 … -0.344892 -0.123227; 0.244099 -0.422044 … -0.463542 0.0235693])
 param([0.582161, -0.0213073, 0.407804, -0.239962, -0.234052, -0.42528, -0.0633082, -0.046206, 0.495791, -1.20858])                                                              
 param([0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0])                                                                                                                       
 param([1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0])                                                                                                                       

julia> testpred = mapleaves(Flux.Tracker.data, testchain)
Chain(Dense(10, 10, NNlib.relu), BatchNorm(10))

julia> testpred(rand(10,1))
10×1 Array{Float64,2}:
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0

julia> Flux.testmode!(testpred)
Chain(Dense(10, 10, NNlib.relu), BatchNorm(10))

julia> testpred(rand(10,1))
10×1 Array{Float64,2}:
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0

julia> params(testchain)
4-element Array{Any,1}:
 param([0.43428 0.419424 … -0.257435 0.165523; 0.169282 -0.0603898 … 0.459877 -0.228518; … ; 0.0777249 0.136552 … -0.344892 -0.123227; 0.244099 -0.422044 … -0.463542 0.0235693])
 param([0.582161, -0.0213073, 0.407804, -0.239962, -0.234052, -0.42528, -0.0633082, -0.046206, 0.495791, -1.20858])                                                              
 param([0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0])                                                                                                                       
 param([1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0])

**Comment by unhandyandy** [comment]
Where is the testmode code located?

**Comment by unhandyandy** [comment]
At line 139 in that file (src/layers/normalise.jl) there is 
`
BN.σ = (1 - mtm) .* BN.σ .+ mtm .* squeeze(data(σ), (axes...)) .* m ./ (m - 1)
`
Do you want to handle the case when m=1?  If m=1 and this code is evaluated then `BN.σ`gets stuck at `Inf`.

**Comment by MikeInnes** [comment]
Sounds like we should just throw an error if `m == 1`? I don't think batchnorm can do anything sensible here when training.

**Comment by unhandyandy** [comment]
Or or you could just pass on the inputs unchanged.

**Comment by MikeInnes** [comment]
Possibly; I'd feel uncomfortable about that unless it's clearly mathematically sensible though, or has precedent as a useful behaviour. Otherwise it just seems like it's making a user error more silent than it needs to be.

**Comment by CarloLucibello** [comment]
not clear what the issue is at this point, closing as inactionable

**Comment by unhandyandy** [comment]
So are you saying the case m=1 never occurs?

**Comment by CarloLucibello** [comment]
You just shouldn't be using batchsize=1 for training models with BatchNorm layers. 
In test phase instead, batchsize=1 is correctly handled.
